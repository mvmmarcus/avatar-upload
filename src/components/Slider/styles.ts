import styled, { css } from "styled-components";

// import * as ButtonStyles from "components/Button/styles";

export const Wrapper = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  max-width: 27.6rem;
`;

export const SliderGroup = styled.div`
  ${({ theme }) => css`
    width: 100%;
    display: flex;
    flex-direction: column;
    max-width: 27.6rem;
    margin-bottom: ${theme.spacings.medium};
  `}
`;

export const Title = styled.span`
  ${({ theme }) => css`
    line-height: 2.8rem;
    font-size: ${theme.font.sizes.medium};
    font-weight: ${theme.font.normal};
    color: ${theme.colors.gray05};
    margin-bottom: ${theme.font.sizes.xsmall};
  `}
`;

export const SliderInput = styled.input.attrs({ type: "range" })`
  ${({ theme, value }) => css`
    width: 100%;
    -webkit-appearance: none;
    -moz-appearance: none;
    outline: 0;
    height: 0.2rem;
    min-height: 0.2rem;
    border-radius: 0.1rem;
    background: ${`linear-gradient(
      to right,
      ${theme.colors.blue01} 0%,
      ${theme.colors.blue01} ${(Number(value) - 1) * 100}%,
      ${theme.colors.blue04} ${(Number(value) - 1) * 100}%,
      ${theme.colors.blue04} 100%
    )`};

    ::-webkit-slider-thumb {
      -webkit-appearance: none;
      width: 12px;
      height: 12px;
      border-radius: 50%;
      background-color: ${theme.colors.blue01};
    }

    ::-moz-range-thumb {
      -moz-appearance: none;
      width: 12px;
      height: 12px;
      border-radius: 50%;
      background-color: ${theme.colors.blue01};
    }
  `}
`;
