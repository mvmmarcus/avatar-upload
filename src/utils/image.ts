type CropImageProps = {
  imageUrl: string;
  cropWidth?: number;
  preserveAspectRatio?: boolean;
  previewImageRatio?: number;
};

const addImageProcess = async (
  imageSource: string
): Promise<HTMLImageElement> => {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.onload = () => resolve(img);
    img.onerror = reject;
    img.src = imageSource;
  });
};

export const cropImage = async ({
  imageUrl,
  cropWidth = 300,
  preserveAspectRatio = true,
  previewImageRatio = 1,
}: CropImageProps) => {
  const image = await addImageProcess(imageUrl);

  const ratio = image.height / image.width;

  const canvas = document.createElement("canvas");
  const context = canvas.getContext("2d");

  if (preserveAspectRatio) {
    canvas.width = image.width < cropWidth ? image.width : cropWidth;
    canvas.height = image.width < cropWidth ? image.height : cropWidth * ratio;

    context?.drawImage(image, 0, 0, canvas.width, canvas.height);
  } else {
    canvas.width = previewImageRatio * image.width;
    canvas.height = canvas.width;

    const cropDistanceX = (image.width - canvas.width) / 2;
    const cropDistanceY = (image.height - canvas.height) / 2;

    context?.drawImage(
      image,
      cropDistanceX,
      cropDistanceY,
      canvas.width,
      canvas.height,
      0,
      0,
      canvas.width,
      canvas.height
    );
  }

  const croppedImageUrl = canvas.toDataURL();

  return croppedImageUrl;
};

export const uploadImage = (file: File): Promise<ProgressEvent<FileReader>> => {
  return new Promise((resolve, reject) => {
    const fileReader = new FileReader();

    fileReader.onload = (event) => resolve(event);
    fileReader.onerror = reject;
    fileReader.readAsDataURL(file);
  });
};
