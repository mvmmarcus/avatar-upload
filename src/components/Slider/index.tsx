import React, { useState } from "react";

import Button from "components/Button";

import { SliderGroup, SliderInput, Title, Wrapper } from "./styles";

export type SliderProps = {
  onValueChange?: (value: number) => void;
  onSave?: (value: number) => void;
};

const Slider = ({ onValueChange, onSave }: SliderProps) => {
  const [value, setValue] = useState(1);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    !!onValueChange && onValueChange(Number(event.target.value));
    setValue(Number(event.target.value));
  };

  const handleSave = (value: number) => {
    !!onSave && onSave(value);
  };
  return (
    <Wrapper>
      <SliderGroup>
        <Title>Crop</Title>
        <SliderInput
          min="1"
          max="2"
          step=".1"
          value={value}
          onChange={handleChange}
          data-testid="slider-input"
        />
      </SliderGroup>
      <Button onClick={() => handleSave(value)}>Save</Button>
    </Wrapper>
  );
};

export default Slider;
