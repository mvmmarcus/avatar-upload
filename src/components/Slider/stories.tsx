import React from "react";

import { Story, Meta } from "@storybook/react/types-6-0";

import Slider, { SliderProps } from ".";

export default {
  title: "Slider",
  component: Slider,
  argTypes: {
    onSave: { action: "saved" },
  },
} as Meta<SliderProps>;

export const Default: Story<SliderProps> = (args) => <Slider {...args} />;

Default.args = {};
