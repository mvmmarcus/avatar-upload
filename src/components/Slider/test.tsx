import React from "react";

import { fireEvent, screen } from "@testing-library/react";

import Slider from ".";
import { renderWithTheme } from "utils/test-utils";

describe("<Slider />", () => {
  it("should render slider as default", () => {
    const { container } = renderWithTheme(<Slider />);

    expect(screen.getByText(/crop/i)).toBeInTheDocument();
    expect(screen.getByTestId("slider-input")).toHaveProperty("value", "1");
    expect(screen.getByRole("button", { name: /save/i })).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should change slider value", () => {
    const onValueChange = jest.fn();

    const { container } = renderWithTheme(
      <Slider onValueChange={onValueChange} />
    );

    const sliderInput = screen.getByTestId("slider-input");

    fireEvent.change(sliderInput, {
      target: {
        value: 1.5,
      },
    });

    expect(onValueChange).toHaveBeenCalledWith(1.5);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should save image crop", () => {
    const onSave = jest.fn();

    const { container } = renderWithTheme(<Slider onSave={onSave} />);

    const sliderInput = screen.getByTestId("slider-input");
    const saveButton = screen.getByRole("button", { name: /save/i });

    fireEvent.change(sliderInput, {
      target: {
        value: 1.5,
      },
    });

    fireEvent.click(saveButton);

    expect(onSave).toHaveBeenCalledWith(1.5);

    expect(container.firstChild).toMatchSnapshot();
  });
});
