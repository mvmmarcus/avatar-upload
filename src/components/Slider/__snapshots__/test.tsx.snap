// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Slider /> should change slider value 1`] = `
.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #fff;
  border: none;
  -webkit-transition: background 0.3s ease-in-out;
  transition: background 0.3s ease-in-out;
  height: 3.2rem;
  font-size: 1.4rem;
  font-weight: 500;
  background-color: #3D485F;
  border: 0.8rem;
  padding: 0.4rem 3.2rem;
  border-radius: 1.6rem;
}

.c4:hover {
  background-color: #677489;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  max-width: 27.6rem;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  max-width: 27.6rem;
  margin-bottom: 3.2rem;
}

.c2 {
  line-height: 2.8rem;
  font-size: 1.6rem;
  font-weight: 400;
  color: #677489;
  margin-bottom: 1.2rem;
}

.c3 {
  width: 100%;
  -webkit-appearance: none;
  -moz-appearance: none;
  outline: 0;
  height: 0.2rem;
  min-height: 0.2rem;
  border-radius: 0.1rem;
  background: linear-gradient( to right, #3F80FF 0%, #3F80FF 50%, #B9D1FF 50%, #B9D1FF 100% );
}

.c3::-webkit-slider-thumb {
  -webkit-appearance: none;
  width: 12px;
  height: 12px;
  border-radius: 50%;
  background-color: #3F80FF;
}

.c3::-moz-range-thumb {
  -moz-appearance: none;
  width: 12px;
  height: 12px;
  border-radius: 50%;
  background-color: #3F80FF;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <span
      class="c2"
    >
      Crop
    </span>
    <input
      class="c3"
      data-testid="slider-input"
      max="2"
      min="1"
      step=".1"
      type="range"
      value="1.5"
    />
  </div>
  <button
    class="c4"
  >
    Save
  </button>
</div>
`;

exports[`<Slider /> should render slider as default 1`] = `
.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #fff;
  border: none;
  -webkit-transition: background 0.3s ease-in-out;
  transition: background 0.3s ease-in-out;
  height: 3.2rem;
  font-size: 1.4rem;
  font-weight: 500;
  background-color: #3D485F;
  border: 0.8rem;
  padding: 0.4rem 3.2rem;
  border-radius: 1.6rem;
}

.c4:hover {
  background-color: #677489;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  max-width: 27.6rem;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  max-width: 27.6rem;
  margin-bottom: 3.2rem;
}

.c2 {
  line-height: 2.8rem;
  font-size: 1.6rem;
  font-weight: 400;
  color: #677489;
  margin-bottom: 1.2rem;
}

.c3 {
  width: 100%;
  -webkit-appearance: none;
  -moz-appearance: none;
  outline: 0;
  height: 0.2rem;
  min-height: 0.2rem;
  border-radius: 0.1rem;
  background: linear-gradient( to right, #3F80FF 0%, #3F80FF 0%, #B9D1FF 0%, #B9D1FF 100% );
}

.c3::-webkit-slider-thumb {
  -webkit-appearance: none;
  width: 12px;
  height: 12px;
  border-radius: 50%;
  background-color: #3F80FF;
}

.c3::-moz-range-thumb {
  -moz-appearance: none;
  width: 12px;
  height: 12px;
  border-radius: 50%;
  background-color: #3F80FF;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <span
      class="c2"
    >
      Crop
    </span>
    <input
      class="c3"
      data-testid="slider-input"
      max="2"
      min="1"
      step=".1"
      type="range"
      value="1"
    />
  </div>
  <button
    class="c4"
  >
    Save
  </button>
</div>
`;

exports[`<Slider /> should save image crop 1`] = `
.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  cursor: pointer;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #fff;
  border: none;
  -webkit-transition: background 0.3s ease-in-out;
  transition: background 0.3s ease-in-out;
  height: 3.2rem;
  font-size: 1.4rem;
  font-weight: 500;
  background-color: #3D485F;
  border: 0.8rem;
  padding: 0.4rem 3.2rem;
  border-radius: 1.6rem;
}

.c4:hover {
  background-color: #677489;
}

.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  max-width: 27.6rem;
}

.c1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  max-width: 27.6rem;
  margin-bottom: 3.2rem;
}

.c2 {
  line-height: 2.8rem;
  font-size: 1.6rem;
  font-weight: 400;
  color: #677489;
  margin-bottom: 1.2rem;
}

.c3 {
  width: 100%;
  -webkit-appearance: none;
  -moz-appearance: none;
  outline: 0;
  height: 0.2rem;
  min-height: 0.2rem;
  border-radius: 0.1rem;
  background: linear-gradient( to right, #3F80FF 0%, #3F80FF 50%, #B9D1FF 50%, #B9D1FF 100% );
}

.c3::-webkit-slider-thumb {
  -webkit-appearance: none;
  width: 12px;
  height: 12px;
  border-radius: 50%;
  background-color: #3F80FF;
}

.c3::-moz-range-thumb {
  -moz-appearance: none;
  width: 12px;
  height: 12px;
  border-radius: 50%;
  background-color: #3F80FF;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <span
      class="c2"
    >
      Crop
    </span>
    <input
      class="c3"
      data-testid="slider-input"
      max="2"
      min="1"
      step=".1"
      type="range"
      value="1.5"
    />
  </div>
  <button
    class="c4"
  >
    Save
  </button>
</div>
`;
